### ISSUE TITLE
An error occurred when using kafkaLineageDispatcher send

### ISSUE THREAD

**Issue by jinmu0410** [AbsaOSS/spline-spark-agent#623]
<img width="625" alt="截屏2023-03-07 21 16 30" src="https://user-images.githubusercontent.com/33084817/223596535-eacdd6ec-75bd-4f13-be34-944b07cfb902.png">

```
Caused by: java.lang.IllegalStateException: Cannot perform operation after producer has been closed
        at: org.apache.kafka.clients.producer.KafkaProducer.throwIfProducerClosed(KafkaProducer.java:863)
```

**Comment by jinmu0410** [comment]
The local mode is normal, but this problem occurs on yarn

**Comment by jinmu0410** [comment]
spark 3.1.2 and yarn 3.1

**Comment by jinmu0410** [comment]
I see that this problem has occurred before, has it been resolved in the future? Now my situation is that this bug is triggered every time

**Comment by glasalvia** [comment]
Is there any news on this topic? Thanks in advance

**Comment by cerveada** [comment]
The issue might be caused by Spark shutting down too quickly and not allowing the agent to finish it's job.

Seems like similar problem happens here: https://github.com/AbsaOSS/spline-spark-agent/issues/478

Could you try to set high `hadoop.service.shutdown.timeout` and let us know if it helps?

**Comment by JinJiDeJinMu** [comment]
I guess it may be the reason why the spark version is packaged？I don't have this problem when I use the spark3.3 version now.

**Comment by jinmu0410** [comment]
I think it may be a problem of compiling and selecting the spark version.
<img width="599" alt="截屏2023-03-25 09 36 26" src="https://user-images.githubusercontent.com/33084817/227678679-41132932-dfde-4f1f-9389-6308317f45f0.png">
this way is wrong

<img width="740" alt="截屏2023-03-25 09 37 08" src="https://user-images.githubusercontent.com/33084817/227678998-18f80b3b-befe-45c9-970a-8c3c1fff97b8.png">
this choice is right.
After choosing the right compilation package, I did not encounter the problem of kafka

**Comment by jinmu0410** [comment]
sorry，This problem occurs again
![截屏2023-03-27 11 17 50](https://user-images.githubusercontent.com/33084817/227832541-9bcc3409-ef4e-40a7-be13-65a19202f9eb.png)

**Comment by cerveada** [comment]
Could you try to set high `hadoop.service.shutdown.timeout` and let us know if it helps?

**Comment by jinmu0410** [comment]
> Could you try to set high `hadoop.service.shutdown.timeout` and let us know if it helps?

I tried it, it doesn't seem to work。

**Comment by jinmu0410** [comment]
I looked at the code and found some key places
<img width="690" alt="截屏2023-03-27 16 09 01" src="https://user-images.githubusercontent.com/33084817/227880886-184d547b-9c9b-41a1-bb4e-4eaf3ad6c191.png">
**sys.addShutdownHook(sf.close())** 
cause **producer.close()**

**Comment by jinmu0410** [comment]
<img width="758" alt="截屏2023-03-27 16 13 04" src="https://user-images.githubusercontent.com/33084817/227881639-a61e0a46-3ad8-4acb-a2a4-7c524aa9afcd.png">
I am now actively calling **sf.close**
@cerveada

**Comment by cerveada** [comment]
What do you mean by _actively calling_ ?

**Comment by jinmu0410** [comment]
After kafka sends the message, then call **SplineRecordSenderFactory.close**

**Comment by cerveada** [comment]
Ok, but then you can't send another message, right?

**Comment by jinmu0410** [comment]
No, I changed some codes to unify plan and event into one kafka message, for some special business later

**Comment by cerveada** [comment]
I found there was an issue in the agent, that cause the Kafka producer being close twice and first time it was too soon. 

The issue: #639 

This issue was fixed in agent version 1.1.0.
